File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if(len(s)==0):
            return 0
        map={}
        maxlen=start=0
        for i in range(len(s)):
            if(s[i] in map and start<=map[s[i]]):
                start=map[s[i]]+1
            else:
                maxlen=max(maxlen,i-start+1)
            map[s[i]]=i
        return maxlen

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=int(s.length());
        int ans=0;
        unordered_map<char,int> map;
        for(int j=0,i=0;j<n;j++){
            if(map[s[j]]>0){//repeated chr
                i=max(map[s[j]],i);
            }
            ans=max(ans,j-i+1);
            map[s[j]]=j+1;
        }
        return ans;
    }
};